<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Leighton Pritchard">

<title>A Little Book of Bioinformatics - MAFFT</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./biological_networks.html" rel="next">
<link href="./multiple_sequence_alignment.html" rel="prev">
<link href="./assets/images/cover.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./multiple_sequence_alignment.html">Multiple Sequence Alignment</a></li><li class="breadcrumb-item"><a href="./mafft.html"><code>MAFFT</code></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">A Little Book of Bioinformatics</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/sipbs-compbiol/little-bioinformatics-book" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface to <em>A Little Book of Bioinformatics</em></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Summary</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text">Representing Biological Data</span></span>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text">Biological Databases</span></span>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text">Pairwise Sequence Alignment</span></span>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./multiple_sequence_alignment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multiple Sequence Alignment</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mafft.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><code>MAFFT</code></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <span class="sidebar-item-text sidebar-link text-start">
 <span class="menu-text">Biological Networks</span></span>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#where-can-i-download-mafft" id="toc-where-can-i-download-mafft" class="nav-link active" data-scroll-target="#where-can-i-download-mafft">Where can I download <code>MAFFT</code>?</a></li>
  <li><a href="#where-can-i-use-mafft-in-a-browser" id="toc-where-can-i-use-mafft-in-a-browser" class="nav-link" data-scroll-target="#where-can-i-use-mafft-in-a-browser">Where can I use <code>MAFFT</code> in a browser?</a></li>
  <li><a href="#how-does-mafft-work" id="toc-how-does-mafft-work" class="nav-link" data-scroll-target="#how-does-mafft-work">How does <code>MAFFT</code> work?</a>
  <ul class="collapse">
  <li><a href="#multiple-workflows-tuned-for-different-circumstances" id="toc-multiple-workflows-tuned-for-different-circumstances" class="nav-link" data-scroll-target="#multiple-workflows-tuned-for-different-circumstances">Multiple workflows, tuned for different circumstances</a></li>
  <li><a href="#pairwise-sequence-or-sequence-group-alignment" id="toc-pairwise-sequence-or-sequence-group-alignment" class="nav-link" data-scroll-target="#pairwise-sequence-or-sequence-group-alignment">Pairwise sequence, or sequence group alignment</a></li>
  <li><a href="#scoring-scheme" id="toc-scoring-scheme" class="nav-link" data-scroll-target="#scoring-scheme">Scoring scheme</a></li>
  <li><a href="#guide-tree-building" id="toc-guide-tree-building" class="nav-link" data-scroll-target="#guide-tree-building">Guide tree building</a></li>
  <li><a href="#rna-and-protein-alignment-incorporating-structural-information" id="toc-rna-and-protein-alignment-incorporating-structural-information" class="nav-link" data-scroll-target="#rna-and-protein-alignment-incorporating-structural-information">RNA and protein alignment incorporating structural information</a></li>
  <li><a href="#extension-of-existing-alignments-with-new-sequences" id="toc-extension-of-existing-alignments-with-new-sequences" class="nav-link" data-scroll-target="#extension-of-existing-alignments-with-new-sequences">Extension of existing alignments with new sequences</a></li>
  <li><a href="#parallelisation" id="toc-parallelisation" class="nav-link" data-scroll-target="#parallelisation">Parallelisation</a></li>
  </ul></li>
  <li><a href="#where-can-i-find-out-more-about-mafft" id="toc-where-can-i-find-out-more-about-mafft" class="nav-link" data-scroll-target="#where-can-i-find-out-more-about-mafft">Where can I find out more about <code>MAFFT</code>?</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/sipbs-compbiol/little-bioinformatics-book/edit/main/mafft.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/sipbs-compbiol/little-bioinformatics-book/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-MAFFT" class="quarto-section-identifier"><code>MAFFT</code></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><code>MAFFT</code> is a very fast progressive alignment method, suitable for aligning large sequence sets, and tuneable for a range of input circumstances.</p>
<p><code>MAFFT</code> stands for Multiple Alignment using Fast Fourier Transform, and was first described in 2002 <span class="citation" data-cites="Katoh2002-nz">(<a href="references.html#ref-Katoh2002-nz" role="doc-biblioref">Katoh et al. 2002</a>)</span>. Since then it has received multiple enhancements and extensions to improve scalability to larger sequence sets, and integration of other data such as protein structures <span class="citation" data-cites="Katoh2008-hj Katoh2013-jm Rozewicki2019-vj Katoh2019-et">(<a href="references.html#ref-Katoh2008-hj" role="doc-biblioref">Katoh and Toh 2008a</a>; <a href="references.html#ref-Katoh2013-jm" role="doc-biblioref">Katoh and Standley 2013</a>; <a href="references.html#ref-Rozewicki2019-vj" role="doc-biblioref">Rozewicki et al. 2019</a>; <a href="references.html#ref-Katoh2019-et" role="doc-biblioref">Katoh, Rozewicki, and Yamada 2019</a>)</span>.</p>
<section id="where-can-i-download-mafft" class="level2">
<h2 class="anchored" data-anchor-id="where-can-i-download-mafft">Where can I download <code>MAFFT</code>?</h2>
<ul>
<li><a href="https://%60MAFFT%60.cbrc.jp/alignment/software/">https://<code>MAFFT</code>.cbrc.jp/alignment/software/</a></li>
</ul>
</section>
<section id="where-can-i-use-mafft-in-a-browser" class="level2">
<h2 class="anchored" data-anchor-id="where-can-i-use-mafft-in-a-browser">Where can I use <code>MAFFT</code> in a browser?</h2>
<ul>
<li><a href="https://mafft.cbrc.jp/alignment/server/">https://mafft.cbrc.jp/alignment/server/</a></li>
<li><a href="https://www.ebi.ac.uk/Tools/msa/%60MAFFT%60/">https://www.ebi.ac.uk/Tools/msa/<code>MAFFT</code>/</a></li>
<li><a href="https://usegalaxy.eu/">https://usegalaxy.eu/</a></li>
</ul>
</section>
<section id="how-does-mafft-work" class="level2">
<h2 class="anchored" data-anchor-id="how-does-mafft-work">How does <code>MAFFT</code> work?</h2>
<section id="multiple-workflows-tuned-for-different-circumstances" class="level3">
<h3 class="anchored" data-anchor-id="multiple-workflows-tuned-for-different-circumstances">Multiple workflows, tuned for different circumstances</h3>
<p><code>MAFFT</code> can be used as a “black box” alignment tool, but it offers several ways to combine its components. The various combinations can be chosen to suit your alignment task or available hardware, and fall into three main categories:</p>
<section id="fft-ns1-fft-ns-2-progressive-methods-for-large-sequence-sets" class="level4">
<h4 class="anchored" data-anchor-id="fft-ns1-fft-ns-2-progressive-methods-for-large-sequence-sets"><code>FFT-NS1</code>, <code>FFT-NS-2</code>: progressive methods for large sequence sets</h4>
<p>These approaches are simple progressive alignment workflows that take advantage of 6-tuple scoring to speed up initial distance matrix construction, and FFTs to speed up sequence and sequence group alignment.</p>
<ol type="1">
<li>A guide tree is constructed from 6-tuple similarity scores</li>
<li>A progressive alignment is performed on the guide tree (and the output alignment returned - <code>FFT-NS-1</code>)</li>
<li>The guide tree is reconstructed from this alignment (<code>FFT-NS-2</code>)</li>
<li>The sequences are realigned against the rebuilt guide tree (and the output alignment returned - <code>FFT-NS-2</code>)</li>
</ol>
<p><code>FFT-NS-2</code> is the default <code>MAFFT</code> algorithm, chosen to balance speed and accuracy <span class="citation" data-cites="Katoh2008-hj">(<a href="references.html#ref-Katoh2008-hj" role="doc-biblioref">Katoh and Toh 2008a</a>)</span>.</p>
</section>
<section id="fft-ns-i-progressive-method-for-more-accurate-alignments-of-smaller-sequence-sets" class="level4">
<h4 class="anchored" data-anchor-id="fft-ns-i-progressive-method-for-more-accurate-alignments-of-smaller-sequence-sets"><code>FFT-NS-i</code>: progressive method for more accurate alignments of smaller sequence sets</h4>
<p>This approach resembles that for <code>FFT-NS-2</code>, but iterates over guide tree reconstruction and sequence realignment until either no improvement can be detected in the alignment (by <em>Weighted Sum of Pairs</em>, WSP), or a maximum number of iterations is reached.</p>
<ol type="1">
<li>A guide tree is constructed from 6-tuple similarity scores</li>
<li>A progressive alignment is performed on the guide tree</li>
<li>The guide tree is reconstructed from this alignment</li>
<li>The sequences are realigned against the rebuilt guide tree</li>
<li>Steps 3 and 4 are repeated until there is no improvement in the alignment’s WSP score, or a maximum number of iterations is reached</li>
</ol>
</section>
<section id="l-ins-i-e-ins-i-and-g-ins-i-consistency-scores-for-difficult-alignment-cases" class="level4">
<h4 class="anchored" data-anchor-id="l-ins-i-e-ins-i-and-g-ins-i-consistency-scores-for-difficult-alignment-cases"><code>L-INS-i</code>, <code>E-INS-i</code>, and <code>G-INS-i</code>: consistency scores for difficult alignment cases</h4>
<p>These approaches resemble <code>FT-NS-i</code> in that they iterate over guide tree construction and realignment, but differ in that they carry out all-vs-all pairwise alignment initially, rather than the 6-tuple approximation, and employ a <code>COFFEE</code>-like <em>consistency score</em> alongside the WSP score, and different alignment algorithms during pairwise alignment.</p>
<ul>
<li><code>E-INS-i</code> uses a global Needleman-Wunsch approach to pairwise alignment, and is suited for alignments between sequences that may have several alignable domains and a high proportion of non-alignable insertions.</li>
<li><code>L-INS-i</code> uses a local Smith-Waterman alignment to improve alignments where there is a single alignable domain, with non-alignable flanking sequence.</li>
<li><code>G-INS-i</code> uses a global Needleman-Wunsch alignment approach, and is suited to situations where it can be assumed that the entire sequence can be aligned for all inputs.</li>
</ul>
</section>
</section>
<section id="pairwise-sequence-or-sequence-group-alignment" class="level3">
<h3 class="anchored" data-anchor-id="pairwise-sequence-or-sequence-group-alignment">Pairwise sequence, or sequence group alignment</h3>
<section id="fast-fourier-transform-speed-up" class="level4">
<h4 class="anchored" data-anchor-id="fast-fourier-transform-speed-up">Fast Fourier Transform speed-up</h4>
<p><code>MAFFT</code> improves on the speed of simpler progressive alignment approaches <span class="citation" data-cites="Feng1987-lz">(<a href="references.html#ref-Feng1987-lz" role="doc-biblioref">Feng and Doolittle 1987</a>)</span> by using efficient algorithms to target slow steps in the process. The main advance, which gives <code>MAFFT</code> its name, was to use the <a href="https://en.wikipedia.org/wiki/Fast_Fourier_transform">fast Fourier Transform (FFT)</a> - an algorithm that converts a signal from one quantitative <em>domain</em>, such as time, to a representation in the <em>frequency domain</em> - to speed up alignment of protein sequences.</p>
<p>A key insight was to note that the frequency of amino acid substitutions is often determined by whether the substituting residue conserves important physicochemical properties <span class="citation" data-cites="Katoh2002-nz">(<a href="references.html#ref-Katoh2002-nz" role="doc-biblioref">Katoh et al. 2002</a>)</span>. Conservation of such properties is a feature of <em>neutral</em> or <em>near-neutral</em> substitutions. In the original paper, each amino acid (<span class="math inline">\(a\)</span>) was assigned to a vector of two values, one representing the amino acid residue’s volume (<span class="math inline">\(v(a)\)</span>), and one representing its polarity (<span class="math inline">\(p(a)\)</span>), normalised to zero mean and unit variance over all amino acids.</p>
<p>By converting the amino acid sequences to be aligned to a sequence of these vectors, <code>MAFFT</code> calculates the <em>correlation</em> between sequences in terms of their polarity and volume components. In doing this, <code>MAFFT</code> considers a <em>positional lag</em> of <span class="math inline">\(k\)</span> sites in the sequence - a parameter controlling the size of sequence region being considered, and the overall correlation is reported for this value, as <span class="math inline">\(c(k)\)</span>. By using a fast Fourier Transform for this operation, the CPU time for a set of <span class="math inline">\(N\)</span> input sequences is reduced from <span class="math inline">\(O(N^2)\)</span> to <span class="math inline">\(O(N log N)\)</span>.</p>
<p>This concept can be extended from an alignment between two sequences to an alignment between two already-aligned groups of sequences, by replacing the vector of volume (or polarity) values for a single sequence with a vector that is the linear combination of the volume components for the aligned group <span class="citation" data-cites="Katoh2002-nz">(<a href="references.html#ref-Katoh2002-nz" role="doc-biblioref">Katoh et al. 2002</a>)</span>.</p>
</section>
<section id="finding-homologous-sequence-segments" class="level4">
<h4 class="anchored" data-anchor-id="finding-homologous-sequence-segments">Finding homologous sequence segments</h4>
<p>If a pair of sequences has homologous regions - conserving polarity and volume - then there will be a corresponding peak in the graph of <span class="math inline">\(c(k)\)</span>. This identifies the <em>positional lag</em> of the match, but not its location. So, to determine the actual location of the match, a <em>sliding window analysis</em> (window size 30) is carried out to identify local homologies for the 20 highest peaks in <span class="math inline">\(c(k)\)</span>. A score threshold of 0.7 per site is applied, and any window with a higher value is considered a <em>homologous segment</em>. Successive segments identified as <em>homologous segments</em> are combined, to a maximum of 150 sites per homolgous segment. Segments longer than this are divided into segments of maximum length 150 <span class="citation" data-cites="Katoh2002-nz">(<a href="references.html#ref-Katoh2002-nz" role="doc-biblioref">Katoh et al. 2002</a>)</span>.</p>
</section>
<section id="aligning-sequence-pairs" class="level4">
<h4 class="anchored" data-anchor-id="aligning-sequence-pairs">Aligning sequence pairs</h4>
<p>A matrix of homologous segments between two sequences is constructed, and the alignment obtained by the standard <a href="https://en.wikipedia.org/wiki/Dynamic_programming"><em>dynamic programming</em></a> procedure <span class="citation" data-cites="Katoh2002-nz">(<a href="references.html#ref-Katoh2002-nz" role="doc-biblioref">Katoh et al. 2002</a>)</span>.</p>
</section>
<section id="application-to-nucleotide-sequences" class="level4">
<h4 class="anchored" data-anchor-id="application-to-nucleotide-sequences">Application to nucleotide sequences</h4>
<p>For nucleotide sequences, rather than using vectors that represent amino acid side chain properties, four-dimensional vectors of A, C, G, T are used at each column. Otherwise the process is identical <span class="citation" data-cites="Katoh2002-nz">(<a href="references.html#ref-Katoh2002-nz" role="doc-biblioref">Katoh et al. 2002</a>)</span>.</p>
</section>
</section>
<section id="scoring-scheme" class="level3">
<h3 class="anchored" data-anchor-id="scoring-scheme">Scoring scheme</h3>
<p><code>MAFFT</code> departs from the common approach of using an all-positive-value scoring matrix, instead adopting a normalised similarity matrix. By default, this was originally derived from a 200 PAM log-odds matrix for both protein and nucleotide alignments, and a simplified gap penalty scheme was also employed <span class="citation" data-cites="Katoh2002-nz">(<a href="references.html#ref-Katoh2002-nz" role="doc-biblioref">Katoh et al. 2002</a>)</span>. More recently, the 200 PAM matrix has been replaced as default by BLOSUM62, and the gap penalty scheme was heavily revised <span class="citation" data-cites="Katoh2008-hj">(<a href="references.html#ref-Katoh2008-hj" role="doc-biblioref">Katoh and Toh 2008a</a>)</span>.</p>
</section>
<section id="guide-tree-building" class="level3">
<h3 class="anchored" data-anchor-id="guide-tree-building">Guide tree building</h3>
<p>The modified UPGMA tree building method used in the original version of <code>MAFFT</code> did not scale well to very large sequence sets, and was replaced in v6 onwards by the PartTree algorithm <span class="citation" data-cites="Katoh2008-hj Katoh2008-if">(<a href="references.html#ref-Katoh2008-hj" role="doc-biblioref">Katoh and Toh 2008a</a>, <a href="references.html#ref-Katoh2008-if" role="doc-biblioref">2008b</a>)</span>. This constructs an approximate tree from unaligned sequences by partitioning the dataset and is <span class="math inline">\(O(N log N)\)</span> rather than <span class="math inline">\(O(N^2)\)</span>, like UPGMA.</p>
</section>
<section id="rna-and-protein-alignment-incorporating-structural-information" class="level3">
<h3 class="anchored" data-anchor-id="rna-and-protein-alignment-incorporating-structural-information">RNA and protein alignment incorporating structural information</h3>
<p>In <code>MAFFT</code> v6, options were introduced to include structural information when aligning RNA sequences. This takes base pairing probability into account, and applies a novel four-way alignment consistency function <span class="citation" data-cites="Katoh2008-hj">(<a href="references.html#ref-Katoh2008-hj" role="doc-biblioref">Katoh and Toh 2008a</a>)</span>.</p>
<p><code>MAFFT</code> v7 introduced the ability to inform an alignment by using the <code>ASH</code> structural alignment program and, in particular, the inter-residue distance between alpha carbons for amino acids in the alignment. This is likely to be most useful when aligning sequence-diverse, but structurally similar, proteins <span class="citation" data-cites="Katoh2013-jm">(<a href="references.html#ref-Katoh2013-jm" role="doc-biblioref">Katoh and Standley 2013</a>)</span>.</p>
<p><code>MAFFT</code> has been integrated with the <code>DASH</code> structural alignment database, to use structural alignment information as a set of constraints on sequence alignment. This integration is available through the <a href="https://mafft.cbrc.jp/alignment/server/">official <code>MAFFT</code> alignment server</a> <span class="citation" data-cites="Rozewicki2019-vj">(<a href="references.html#ref-Rozewicki2019-vj" role="doc-biblioref">Rozewicki et al. 2019</a>)</span>.</p>
</section>
<section id="extension-of-existing-alignments-with-new-sequences" class="level3">
<h3 class="anchored" data-anchor-id="extension-of-existing-alignments-with-new-sequences">Extension of existing alignments with new sequences</h3>
<p>Further improvements were introduced in <code>MAFFT</code> v7 to enable addition of unaligned sequences into an existing multiple sequence alignment <span class="citation" data-cites="Katoh2013-jm">(<a href="references.html#ref-Katoh2013-jm" role="doc-biblioref">Katoh and Standley 2013</a>)</span>.</p>
</section>
<section id="parallelisation" class="level3">
<h3 class="anchored" data-anchor-id="parallelisation">Parallelisation</h3>
<p>Threading was added to <code>MAFFT</code> v7. Progressive alignment method outputs are identical in non-threaded and threaded runs, but iterative refinement-based alignments may generate different output when threaded <span class="citation" data-cites="Katoh2013-jm">(<a href="references.html#ref-Katoh2013-jm" role="doc-biblioref">Katoh and Standley 2013</a>)</span>.</p>
</section>
</section>
<section id="where-can-i-find-out-more-about-mafft" class="level2">
<h2 class="anchored" data-anchor-id="where-can-i-find-out-more-about-mafft">Where can I find out more about <code>MAFFT</code>?</h2>
<ul>
<li><a href="https://mafft.cbrc.jp/alignment/software/tips0.html"><code>MAFFT</code> v7 tips</a></li>
<li><a href="https://mafft.cbrc.jp/alignment/software/algorithms/algorithms.html"><code>MAFFT</code> v7 algorithm</a></li>
</ul>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-Feng1987-lz" class="csl-entry" role="listitem">
Feng, Da-Fei, and Russell F Doolittle. 1987. <span>“Progressive Sequence Alignment as a Prerequisite to Correct Phylogenetic Trees.”</span> <em>J. Mol. Evol.</em> 25 (4): 351–60. <a href="https://doi.org/10.1007/BF02603120">https://doi.org/10.1007/BF02603120</a>.
</div>
<div id="ref-Katoh2002-nz" class="csl-entry" role="listitem">
Katoh, Kazutaka, Kazuharu Misawa, Kei-Ichi Kuma, and Takashi Miyata. 2002. <span>“<span>MAFFT</span>: A Novel Method for Rapid Multiple Sequence Alignment Based on Fast Fourier Transform.”</span> <em>Nucleic Acids Res.</em> 30 (14): 3059–66. <a href="https://doi.org/10.1093/nar/gkf436">https://doi.org/10.1093/nar/gkf436</a>.
</div>
<div id="ref-Katoh2019-et" class="csl-entry" role="listitem">
Katoh, Kazutaka, John Rozewicki, and Kazunori D Yamada. 2019. <span>“<span>MAFFT</span> Online Service: Multiple Sequence Alignment, Interactive Sequence Choice and Visualization.”</span> <em>Brief. Bioinform.</em> 20 (4): 1160–66. <a href="https://doi.org/10.1093/bib/bbx108">https://doi.org/10.1093/bib/bbx108</a>.
</div>
<div id="ref-Katoh2013-jm" class="csl-entry" role="listitem">
Katoh, Kazutaka, and Daron M Standley. 2013. <span>“<span>MAFFT</span> Multiple Sequence Alignment Software Version 7: Improvements in Performance and Usability.”</span> <em>Mol. Biol. Evol.</em> 30 (4): 772–80. <a href="https://doi.org/10.1093/molbev/mst010">https://doi.org/10.1093/molbev/mst010</a>.
</div>
<div id="ref-Katoh2008-hj" class="csl-entry" role="listitem">
Katoh, Kazutaka, and Hiroyuki Toh. 2008a. <span>“Recent Developments in the <span>MAFFT</span> Multiple Sequence Alignment Program.”</span> <em>Brief. Bioinform.</em> 9 (4): 286–98. <a href="https://doi.org/10.1093/bib/bbn013">https://doi.org/10.1093/bib/bbn013</a>.
</div>
<div id="ref-Katoh2008-if" class="csl-entry" role="listitem">
———. 2008b. <span>“Recent Developments in the <span>MAFFT</span> Multiple Sequence Alignment Program.”</span> <em>Brief. Bioinform.</em> 9 (4): 286–98. <a href="https://doi.org/10.1093/bib/bbn013">https://doi.org/10.1093/bib/bbn013</a>.
</div>
<div id="ref-Rozewicki2019-vj" class="csl-entry" role="listitem">
Rozewicki, John, Songling Li, Karlou Mar Amada, Daron M Standley, and Kazutaka Katoh. 2019. <span>“<span>MAFFT-DASH</span>: Integrated Protein Sequence and Structural Alignment.”</span> <em>Nucleic Acids Res.</em> 47 (W1): W5–10. <a href="https://doi.org/10.1093/nar/gkz342">https://doi.org/10.1093/nar/gkz342</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./multiple_sequence_alignment.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Multiple Sequence Alignment</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./biological_networks.html" class="pagination-link">
        <span class="nav-page-text">Biological Networks</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">This book was written by Leighton Pritchard.</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://quarto.org/">Quarto</a>.<br>
</div>
  </div>
</footer>



</body></html>